src/player/systems/attacks.rs

```diff
@@ -18,90 +24,82 @@ pub fn player_scale_fire_rate_system(mut player_query: Query<&mut PlayerComponen
 }
 
 /// Manages the players firing weapons
-#[allow(clippy::too_many_arguments)]
 pub fn player_fire_weapon_system(
-    gamepads: Res<Gamepads>,
-    gamepad_input: Res<Input<GamepadButton>>,
-    mouse_input: Res<Input<MouseButton>>,
-    keyboard_input: Res<Input<KeyCode>>,
-    mut player_query: Query<(&mut PlayerComponent, &Velocity, &Transform, Entity)>,
+    mut player_query: Query<(
+        &mut PlayerComponent,
+        &Velocity,
+        &Transform,
+        &ActionState<PlayerAction>,
+        Entity,
+    )>,
     time: Res<Time>,
     mut spawn_projectile: EventWriter<SpawnProjectileEvent>,
     mut sound_effect_event_writer: EventWriter<PlaySoundEffectEvent>,
-    players_resource: Res<PlayersResource>,
+    run_resource: Res<CurrentRunProgressResource>,
 ) {
-    // get keyboard fire input
-    let fire_keyboard_input =
-        keyboard_input.pressed(KeyCode::Space) || mouse_input.pressed(MouseButton::Left);
-
-    // get gamepad fire input
-    let fire_gamepad_inputs: HashMap<usize, bool> = gamepads
-        .iter()
-        .map(|gamepad| {
-            (
-                gamepad.id,
-                gamepad_input.pressed(GamepadButton {
-                    gamepad,
-                    button_type: GamepadButtonType::RightTrigger,
-                }),
-            )
+    // first make sure that the players can shoot based on the level
+    let can_fire_weapon = run_resource.current_level.clone().is_some_and(|level| {
```

Comment:
I think we can avoid cloning this by taking a reference somewhere (probably like `|&level|` and `if let ... = &thing) `